starport/services/chain/sourcetracker.go

```diff
@@ -0,0 +1,302 @@
+package chain
+
+import (
+	"context"
+	"fmt"
+	"os/exec"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	moduleanalysis "github.com/tendermint/starport/starport/pkg/cosmosanalysis/module"
+	"github.com/tendermint/starport/starport/pkg/dirchange"
+)
+
+const (
+	moduleWildcard = "*"
+)
+
+type ArtifactType uint8
+
+const (
+	ArtifactTypeBinary = iota
+	ArtifactTypeDart
+	ArtifactTypeOpenAPI
+	ArtifactTypeJS
+	ArtifactTypeModuleProto
+	// ArtifactTypeState represents app state created from genesis
+	ArtifactTypeState
+)
+
+// Artifact represents a concrete building process output e.g. module: gov artifactType: ArtifactTypeDart
+type Artifact struct {
+	module       string
+	artifactType ArtifactType
+}
+
+type Artifacts map[Artifact]struct{}
+
+func newArtifacts() Artifacts {
+	return Artifacts{}
+}
+
+func (a Artifacts) addArtifacts(artifacts Artifacts) {
+	for artifact := range artifacts {
+		a[artifact] = struct{}{}
+	}
+}
+
+// Contains checks whether Artifacts contain specific artifact of not
+func (a Artifacts) Contains(module string, artifactType ArtifactType) bool {
+	_, specificArtifactExists := a[Artifact{
+		module:       module,
+		artifactType: artifactType,
+	}]
+
+	_, wildcardArtifactExists := a[Artifact{
+		module:       "",
+		artifactType: artifactType,
+	}]
+
+	return specificArtifactExists || wildcardArtifactExists
+}
+
+type TrackerType uint8
+
+const (
+	GoSourceTracker = iota
+	BinaryTracker
+	ModuleProtoSourceTracker
+	VuexOutputTracker
+	ModuleProtoGoOutputTracker
+	OpenApiOutputTracker
+	DartOutputTracker
+	ConfigTracker
+	ThirdPartyProtoTracker
+	GoModTracker
+)
+
+// Tracker tracks specified file changes for specified module set. Empty module slice represents all modules at once
+type Tracker struct {
+	paths       []string
+	modules     []string
+	trackerType TrackerType
+}
+
+// trackerConfig specifies which artifacts derive from different tracker types
+var trackerConfig = map[TrackerType][]ArtifactType{
+	GoSourceTracker:            {ArtifactTypeBinary},
+	BinaryTracker:              {ArtifactTypeBinary},
+	ModuleProtoSourceTracker:   {ArtifactTypeModuleProto, ArtifactTypeDart, ArtifactTypeJS, ArtifactTypeOpenAPI, ArtifactTypeBinary},
+	VuexOutputTracker:          {ArtifactTypeJS},
+	ModuleProtoGoOutputTracker: {ArtifactTypeModuleProto, ArtifactTypeBinary},
+	OpenApiOutputTracker:       {ArtifactTypeOpenAPI},
+	DartOutputTracker:          {ArtifactTypeDart},
+	ConfigTracker:              {ArtifactTypeState},
+	ThirdPartyProtoTracker:     {ArtifactTypeModuleProto, ArtifactTypeDart, ArtifactTypeJS, ArtifactTypeOpenAPI, ArtifactTypeBinary},
+	GoModTracker:               {ArtifactTypeModuleProto, ArtifactTypeDart, ArtifactTypeJS, ArtifactTypeOpenAPI, ArtifactTypeBinary},
+}
+
+func (t Tracker) getTargetArtifacts() Artifacts {
+	// using trackerConfig gets list of artifacts derived from the tracker
+	return nil
+}
+
+// isChanged checks either tracked file were changed since last persist or not
```

Comment:
```suggestion
// isChanged checks if tracked files were changed since last persist
```

Reply:
when we see the word "either" that is an opportunity to be concise with language, here it's a boolean so we can say it tracks changes yes (so the default is yes)

we can usually just omit the "or not"